### PR TITLE
Fix anomalies in Deposits in the Conway Era

### LINKED ISSUE #3834: Fix anomalies in Deposits in the Conway Era.
Code for handling deposits in the Conway Era has at least 3 oversights

1) getTotalDepositsTxBody ignores DRep deposits
2) obligationCertState also ignores DRep deposits
3) The Assertion violations 
     a)  "Deposit pot must equal obligation (PoolReap)"
     b)  "Deposit pot must equal obligation (Ledger)"
     Do not even tell what the pot and obligation is, so its is very hard to diagnose what is wrong
4) obligationCertState, instead of returning a single Coin should return a Triple
        CertObligation { drep :: Coin, pool::Coin, stake:: Coin}
5)  totalObligation, instead of returning a single Coin should return a Quadruple

### PR DESCRIPTION
**Pull request by TimSheard**
Fixes #3834 
# Checklist

- [x] Commit sequence broadly makes sense and commits have useful messages
- [x] New tests are added if needed and existing tests are updated
- [ ] When applicable, versions are updated in `.cabal` and `CHANGELOG.md` files according to the
      [versioning process](https://github.com/input-output-hk/cardano-ledger/blob/master/RELEASING.md#versioning-process).
- [ ] The version bounds in `.cabal` files for all affected packages are updated. **If you change the bounds in a cabal file, that package itself must have a version increase.** (See [RELEASING.md](https://github.com/input-output-hk/cardano-ledger/blob/master/RELEASING.md#versioning-process))
- [ ] All visible changes are prepended to the latest section of a `CHANGELOG.md` for the affected packages. **New section is never added with the code changes.** (See [RELEASING.md](https://github.com/input-output-hk/cardano-ledger/blob/master/RELEASING.md#changelogmd))
- [x] Code is formatted with [`fourmolu`](https://github.com/fourmolu/fourmolu) (use `scripts/fourmolize.sh`)
- [x] Cabal files are formatted (use `scripts/cabal-format.sh`)
- [x] [`hie.yaml`](https://github.com/input-output-hk/cardano-ledger/blob/master/hie.yaml) has been updated (use `scripts/gen-hie.sh`)
- [ ] Self-reviewed the diff

### CODE REVIEW

File: libs/cardano-ledger-core/src/Cardano/Ledger/CertState.hs [hunk 1]
```diff
@@ -461,12 +464,58 @@ refundPoolDeposit keyhash pstate = (coin, pstate {psDeposits = newpool})
       Just c -> (c, Map.delete keyhash pool)
       Nothing -> (mempty, pool)
 
+refundDRepDeposit :: Credential 'DRepRole (EraCrypto era) -> VState era -> (Coin, VState era)
+refundDRepDeposit cred vstate = (coin, newVState)
+  where
+    (coin, newVState) = case Map.lookup cred (vsDReps vstate) of
+      Just drepState -> (drepDeposit drepState, vstate {vsDReps = Map.delete cred (vsDReps vstate)})
+      Nothing -> (mempty, vstate)
```
Comment (lehins) [inline libs/cardano-ledger-core/src/Cardano/Ledger/CertState.hs]: This function is not used anywhere

```suggestion
```

File: libs/cardano-ledger-core/src/Cardano/Ledger/CertState.hs [hunk 2]
```diff
@@ -435,7 +438,7 @@ ptrsMap :: DState era -> Map Ptr (Credential 'Staking (EraCrypto era))
 ptrsMap (DState {dsUnified = UMap _ ptrmap}) = ptrmap
 
 -- ==========================================================
--- Functions that handle Deposits for pool key hashes.
+-- Functions that handling Deposits
```
Comment (lehins) [inline libs/cardano-ledger-core/src/Cardano/Ledger/CertState.hs]: Is "handling" in a correct form for this sentence?
```suggestion
-- Functions that handle Deposits
```

**Review by lehins — APPROVED** [review]
Looks great. Thank you

**Review by teodanciu — APPROVED** [review]
Interesting, so the reason the assertion wasn't failing is because the value happened to be missing from both sides of the equation I suppose.  
Looks good to me.

File: libs/cardano-ledger-core/src/Cardano/Ledger/CertState.hs (hunk 1 above)
Comment (TimSheard) [reply libs/cardano-ledger-core/src/Cardano/Ledger/CertState.hs]: done

File: libs/cardano-ledger-core/src/Cardano/Ledger/CertState.hs (hunk 2 above)
Comment (TimSheard) [reply libs/cardano-ledger-core/src/Cardano/Ledger/CertState.hs]: done

**Pull request merged**